# 00. Table of contents
 - Importing libraries
 - Importing Dataset
 - Looking for mixed type data and addressing them
 - Missing values and addressing them
 - Duplicates and addressing them

# 01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing Dataset


In [2]:
path = r'C:\Users\viki\Documents\Data Analytics\Immersion\Achievement 4\Instacart Basket Analysis'


In [3]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

C:\Users\viki\anaconda3\lib\site-packages\numpy\lib\arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Test Exercise

## 03.a - Mixed type data

In [5]:
# creating a new df
df_test = pd.DataFrame()

In [6]:
#adding new column w/ mixed values
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# change datatype to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_test['mix'].dtype

dtype('O')

## 03.b - Missing values

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#creating a subset of rows with missing values
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
#checking number of rows
df_prods.shape

(49693, 5)

In [15]:
# removing rows with NAs
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

## 03.c - Duplicates

In [17]:
#looking for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
#checking No. of rows
df_prods_clean.shape

(49677, 5)

In [20]:
#dropping duplicates by creating new df
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
#checking No.of rows in new df
df_prods_clean_no_dups.shape

(49672, 5)

# 04. Task

## Step 2 
### Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. 

In [22]:
df_prods.describe().round(2)

product_id  aisle_id  department_id    prices
count    49693.00  49693.00       49693.00  49693.00
mean     24844.35     67.77          11.73      9.99
std      14343.72     38.32           5.85    453.52
min          1.00      1.00           1.00      1.00
25%      12423.00     35.00           7.00      4.10
50%      24845.00     69.00          13.00      7.10
75%      37265.00    100.00          17.00     11.20
max      49688.00    134.00          21.00  99999.00

<font color=blue>The maximum price of 99999 seems unreasonal, this could be further investigated</font> 

## Step 3
### Check for mixed-type data in your df_ords dataframe.

In [23]:
for col in df_ords.columns.tolist():
  weird2 = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird2]) > 0:
    print (col)

<font color=blue>The command didnt find any mixed type data. QUESTION: if I used earlier in the exercise the variable 'weird' than here it is better to use a new name for the variable (weird2), right? does it have a significance ot it would be okay to use here "weird" again as I am running the two cells separately anyway?</font> 

## Step 4 
### If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

<font color=blue>There were none</font> 

## Step 5 
### Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [24]:
#looking for missing values
df_ords.isnull().sum() 

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [25]:
#checking No. of rows in the df
df_ords.shape

(3421083, 6)

In [26]:
#checking frequency on the column with missing values
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

<font color=blue>Approximately 6% of the values are missing in the "days since prior order" column.The reason for the missing value could be that that order was the very first of the customer so the value should be 0. But we see in the fequency table that we have 0 values in our column, so that is probably not the real reason why it is missing. It could be just randomly missing by forgetting to record the value during the data collection</font> 

## Step 6 
### Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [27]:
#looking for mean
df_ords['days_since_prior_order'].mean()

11.114836226863012

In [28]:
#imputing mean (11) for the missing values
df_ords_clean = df_ords.copy()
df_ords_clean['days_since_prior_order'] = df_ords['days_since_prior_order'].fillna(11)
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                    11.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [29]:
#checking if NaNs got imputed in the new df
df_ords_clean['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
11.0    287179
6.0     240013
4.0     221696
3.0     217005
5.0     214503
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [30]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                    11.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

<font color=blue>As there were no extreme outliers, I imputed the missing values with the mean
QUESTION: my script has an issue: it overwrites also the NaNs in the original df and this is not what I wanted.I wanted to create this clean version and leave the original unchanged. What should I do differently to achieve this?</font> 

## Step 7 
### Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.


In [31]:
#finding duplicates
df_dups2 = df_ords_clean[df_ords_clean.duplicated()]

In [32]:
#viewing duplicates
df_dups2

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

<font color=blue>There are no duplicates</font> 

## Step 8  
### Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.


<font color=blue>There are no duplicates</font>

## Step 9 
### Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [33]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned_final.csv')) #exporting df_ords_clean

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned_final.csv')) #exporting df_ords_clean_no_dups

In [35]:
df_ords_clean.shape


(3421083, 6)

In [38]:
df_prods_clean_no_dups.shape

(49672, 5)